src/govuk/components/exit-this-page/exit-this-page.test.js

```diff
@@ -2,38 +2,133 @@
+
+      await Promise.all([
+        page.waitForNavigation(),
+        page.click(buttonClass)
+      ])
+
+      const url = await page.url()
+      expect(url).toBe(href)
+    })
+
+    it('navigates to the href of the skiplink', async () => {
+      await goToExample(page, 'exit-this-page-with-skiplink')
+
+      const href = await page.evaluate((skiplinkClass) => document.querySelector(skiplinkClass).href, skiplinkClass)
+
+      await Promise.all([
+        page.waitForNavigation(),
+        page.focus(skiplinkClass),
+        page.click(skiplinkClass)
+      ])
+
+      const url = await page.url()
+      expect(url).toBe(href)
+    })
+
+    it('activates the button functionality when the Shift key is pressed 3 times', async () => {
+      await goToComponent(page, 'exit-this-page')
+
+      const href = await page.evaluate((buttonClass) => document.querySelector(buttonClass).href, buttonClass)
+
+      await Promise.all([
+        page.keyboard.press('Shift'),
+        page.keyboard.press('Shift'),
+        page.keyboard.press('Shift'),
+        page.waitForNavigation()
+      ])
+
+      const url = await page.url()
+      expect(url).toBe(href)
+    })
+
+    it('shows the ghost page when the EtP button is clicked', async () => {
+      await goToExample(page, 'exit-this-page-with-skiplink')
+
+      // Make the button not navigate away from the current page
+      await page.evaluate((buttonClass) => document.body.querySelector(buttonClass).setAttribute('href', '#'), buttonClass)
+      await page.click(buttonClass)
+
+      const ghostOverlay = await page.evaluate((overlayClass) => document.body.querySelector(overlayClass), overlayClass)
+      expect(ghostOverlay).not.toBeNull()
+    })
+
+    it('shows the ghost page when the skiplink is clicked', async () => {
+      await goToExample(page, 'exit-this-page-with-skiplink')
+
+      // Make the button not navigate away from the current page
+      await page.evaluate((skiplinkClass) => document.body.querySelector(skiplinkClass).setAttribute('href', '#'), skiplinkClass)
+      await page.focus(skiplinkClass)
+      await page.click(skiplinkClass)
+
+      const ghostOverlay = await page.evaluate((overlayClass) => document.body.querySelector(overlayClass), overlayClass)
+      expect(ghostOverlay).not.toBeNull()
+    })
+
+    it('hides the ghost page when the user navigates back in history', async () => {
+      await goToExample(page, 'exit-this-page-with-skiplink')
+
+      await Promise.all([
+        page.waitForNavigation(),
+        page.click(buttonClass)
+      ])
 
-    await Promise.all([
-      page.keyboard.press('Shift'),
-      page.keyboard.press('Shift'),
-      page.keyboard.press('Shift'),
-      page.waitForNavigation()
-    ])
+      await page.goBack()
 
-    const url = await page.url()
-    expect(url).toBe(href)
+      const ghostOverlay = await page.evaluate((overlayClass) => document.body.querySelector(overlayClass), overlayClass)
```

Comment:
I'm not sure this has done the trick sadly. When I remove the contents of `syncOverlayVisibility` so that the overlay isn't removed this still reports `null`. I had a play around but I couldn't figure it out. It's returning the page successfully, just not retaining the overlay.

Don't feel that you have to solve this right now. I suspect this is something we need to figure out in our own testing infrastructure.

Reply:
I think I have a new theory here: Puppeteer runs Chrome tests in incognito mode. Does that mode avoid using bfcache at all, always loaded anew rather than restoring the previous JS state? This would result in it passing the test without needing to run the actual function.

Alas I'm struggling to find anything concrete on this on t'interweb. If there are any weird interactions between private browsing and bfcache, it doesn't seem to be particularly well documented.